File: solutions/binary-search/binary-search.cpp
// @Title: Binary Search
// @Author: colinjxc
// @Date: 2022-01-10T18:44:34+08:00
// @URL: https://leetcode-cn.com/problems/binary-search


class Solution {
public:
    int search(vector<int>& nums, int target) {
        int l = 0, r = nums.size() - 1;
        while (l <= r) {
            int mid = l + (r - l)/2;
            if (target > nums[mid]) {
                l = mid + 1;
            } else if (target < nums[mid]) {
                r = mid - 1;
            } else {
                return mid;
            }
        }
        return -1;
    }
};

File: solutions/binary-search/binary-search.rs
// @Title: Binary Search
// @Author: colinjxc
// @Date: 2021-10-22T09:40:58+08:00
// @URL: https://leetcode-cn.com/problems/binary-search


impl Solution {
    pub fn search(nums: Vec<i32>, target: i32) -> i32 {
        let mut left = 0;
        let mut right = nums.len();
        while left < right {
            let mid = left + ((right - left) >> 1) ;
            if target == nums[mid] {
                return mid as i32;
            } else if target < nums[mid] {
                right = mid;
            } else {
                left = mid + 1;
            }
        }

        return -1;
    }
}
